### ISSUE TITLE
New MSTest runner doesn' t integrate with coverlet

### ISSUE THREAD

**Issue by MarkusRodler** [microsoft/testfx#3189]
Previously I used `dotnet test /p:CollectCoverage=true` with `coverlet.msbuild` to produce an coverage output on the command line like that:
```sh
+--------+------+--------+--------+
| Module | Line | Branch | Method |
+--------+------+--------+--------+
| Maui   | 0%   | 100%   | 0%     |
+--------+------+--------+--------+
| Shared | 0%   | 0%     | 0%     |
+--------+------+--------+--------+
| Client | 0%   | 100%   | 0%     |
+--------+------+--------+--------+
| Test   | 0%   | 0%     | 0%     |
+--------+------+--------+--------+

+---------+------+--------+--------+
|         | Line | Branch | Method |
+---------+------+--------+--------+
| Total   | 0%   | 0%     | 0%     |
+---------+------+--------+--------+
| Average | 0%   | 50%    | 0%     |
+---------+------+--------+--------+
```
But with the new MS Test runner this isn't possible anymore. It would be nice that this feature is restored. Or if someone has an alternative that shows the coverage for each module.

**Comment by MarcoRossignoli** [comment]
The new testing platform upon which the MSTest runner is build has got different extensibility point and MSBuild integration. 
You can open an issue on coverlet repo for it https://github.com/coverlet-coverage/coverlet/issues
You can use the coverlet tool for it https://github.com/coverlet-coverage/coverlet?tab=readme-ov-file#net-global-tool-guide-suffers-from-possible-known-issue or move to the MS coverage version that's already integrated https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-extensions-code-coverage#microsoft-code-coverage
With the MSTest.Sdk trx and codecoverage are already opted in by default https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-sdk#mstest-runner-profile

**Comment by MarkusRodler** [comment]
Thanks for the reply!

> You can use the coverlet tool for now https://github.com/coverlet-coverage/coverlet?tab=readme-ov-file#net-global-tool-guide-suffers-from-possible-known-issue 

That works

> or move to the MS coverage version that's already integrated https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-extensions-code-coverage#microsoft-code-coverage
 
That works but does not give the coverage output on the command line.

> With the MSTest.Sdk trx and codecoverage are already opted in by default https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-sdk#mstest-runner-profile

I think TRX is not useful for this scenario.

**Comment by MarcoRossignoli** [comment]
> That works but does not give the coverage output on the command line.

Only the msbuild integration scenario of Coverlet shows information in console, usually ppl that are not using it are generating a report with tools like report generator https://github.com/danielpalme/ReportGenerator

**Comment by MarkusRodler** [comment]
I find the coverage output on the command line very useful. It works with the coverlet tool but as far as I understand you have to first run the tests with dotnet run to generate a dll. And then you can run the coverlet tool with that dll. This means that I can't run the tool with dotnet watch to always run the tests and output the current coverage. Even not if I make a msbuild task for that because msbuild doesn't work with dotnet watch.

**Comment by MarcoRossignoli** [comment]
Console output is useful but doesn't scale for big projects, usually IDEs offers contextual report like Visual Studio as explained here https://devblogs.microsoft.com/visualstudio/code-coverage-features-in-visual-studio-enterprise/

>  This means that I can't run the tool with dotnet watch to always run the tests and output the current coverage.

Are you using the hot reload extension https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-extensions-hosting#hot-reload?

**Comment by MarkusRodler** [comment]
Yes that doesn't scale for big projects but we use microservices and that fits better.

> Are you using the hot reload extension https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-extensions-hosting#hot-reload?

Not yet

**Comment by connorjs** [comment]
> With the MSTest.Sdk trx and codecoverage are already opted in by default https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-sdk#mstest-runner-profile
Here the documentation https://github.com/microsoft/codecoverage/blob/main/samples/Algorithms/scenarios/scenario07/README.md

How does this work with running multiple test projects in a solution? Before using the MSTest Runner, which I'm trying with MSTest.Sdk, I would use `dotnet test` with a run settings file or command line arguments.

I have struggled to find the "solution wide coverage" approach. I'd be happy with multiple files, one per project, in a central directory. Merging is trivial with tools like report generator (which ADO pipelines use).

**Comment by MarcoRossignoli** [comment]
> I have struggled to find the "solution wide coverage" approach.

If you've a solution with all MSTest runner configuration(`MStest.Sdk`) there's no automatic way to merge at solution level with `dotnet test`. We've planned to support it.
So the solution is to merge at the end in a custom step using the `dotnet-coverage tool` (`dotnet-coverage merge`), here you can find some sample https://github.com/microsoft/codecoverage/blob/main/samples/Calculator/scenarios/scenario18/README.md

**Comment by MarkusRodler** [comment]
@Evangelink this isn't a duplicate. In the referenced issue is no sentence about the missing coverage output which is the point of this bug issue.

**Comment by Evangelink** [comment]
This is a duplicate because coverlet extension works only for VSTest leading to coverage not being produced. We need coverlet to implement an extension to the new platform or you need to update your command for it to work.
